pkg/morph/deploy/util.go

```diff
@@ -166,3 +187,136 @@ func readNNSOnChainState(b Blockchain) (*state.Contract, error) {
+func (x contractVersion) cmp(y contractVersion) int {
+	xN := x.major*majorSpace + x.minor*minorSpace + x.patch
+	yN := y.major*majorSpace + y.minor*minorSpace + y.patch
+	if xN < yN {
+		return -1
+	} else if xN == yN {
+		return 0
+	}
+	return 1
+}
+
+func (x contractVersion) String() string {
+	const sep = "."
+	return fmt.Sprintf("%d%s%d%s%d", x.major, sep, x.minor, sep, x.patch)
+}
+
+// parses contractVersion from the invocation result of methodVersion method.
+func parseContractVersionFromInvocationResult(res *result.Invoke) (contractVersion, error) {
+	bigVersionOnChain, err := unwrap.BigInt(res, nil)
+	if err != nil {
+		return contractVersion{}, fmt.Errorf("unwrap big integer from '%s' method return: %w", methodVersion, err)
+	} else if !bigVersionOnChain.IsUint64() {
+		return contractVersion{}, fmt.Errorf("invalid/unsupported format of the '%s' method return: expected uint64, got %v", methodVersion, bigVersionOnChain)
+	}
+
+	n := bigVersionOnChain.Uint64()
+
+	mjr := n / majorSpace
+
+	return contractVersion{
+		major: mjr,
+		minor: (n - mjr*majorSpace) / minorSpace,
+		patch: n % minorSpace,
+	}, nil
+}
+
+// readContractOnChainVersion returns current version of the smart contract
+// presented in given Blockchain with specified address.
+func readContractOnChainVersion(b Blockchain, onChainAddress util.Uint160) (contractVersion, error) {
+	res, err := invoker.New(b, nil).Call(onChainAddress, methodVersion)
+	if err != nil {
+		return contractVersion{}, fmt.Errorf("call '%s' contract method: %w", methodVersion, err)
+	}
+
+	return parseContractVersionFromInvocationResult(res)
+}
+
+// readContractLocalVersion returns current version of the smart contract
+// presented in given Blockchain with specified address.
+func readContractLocalVersion(rpc invoker.RPCInvoke, localNEF nef.File, localManifest manifest.Manifest) (contractVersion, error) {
+	jManifest, err := json.Marshal(localManifest)
+	if err != nil {
+		return contractVersion{}, fmt.Errorf("encode manifest into JSON: %w", err)
+	}
+
+	bNEF, err := localNEF.Bytes()
+	if err != nil {
+		return contractVersion{}, fmt.Errorf("encode NEF into binary: %w", err)
+	}
+
+	script := io.NewBufBinWriter()
+	emit.Opcodes(script.BinWriter, opcode.NEWARRAY0)
+	emit.Int(script.BinWriter, int64(callflag.All))
+	emit.String(script.BinWriter, methodVersion)
+	emit.AppCall(script.BinWriter, management.Hash, "deploy", callflag.All, bNEF, jManifest)
+	emit.Opcodes(script.BinWriter, opcode.PUSH2, opcode.PICKITEM)
+	emit.Syscall(script.BinWriter, interopnames.SystemContractCall)
+
+	res, err := invoker.New(rpc, nil).Run(script.Bytes())
+	if err != nil {
+		return contractVersion{}, fmt.Errorf("run test script deploying contract and calling its '%s' method: %w", methodVersion, err)
+	}
+
+	return parseContractVersionFromInvocationResult(res)
+}
+
+type transactionGroupWaiter interface {
+	WaitAny(ctx context.Context, vub uint32, hashes ...util.Uint256) (*state.AppExecResult, error)
+}
```

Comment:
my goal was to make it possible to pass `notary.Actor` as waiter. Notary actor doesn't implement `actor.Waiter` (another `Wait` signature). So we need either make impossible to use `notary.Actor` as waiter or keep this narrowed in-place interface.

Reply:
Reasonable, and since we'd like to keep notary Actor's `Wait` as is (because it's needed for the users), then I doubt we can make Notary Actor implement Waiter. So let's keep it as is.